Add Quiltctl

kklin opened the pull request:
This series divides Quilt into a daemon and a client. The daemon is responsible for enforcing Stitches. The client sends Stitches to the daemon to be enforced. It is also capable of querying the state of the daemon, getting imports, etc.

mmussomele commented on Makefile:
```diff
@@ -61,7 +66,7 @@ format-check:
 	cd -P . && go vet $(PACKAGES)
 	for package in $(PACKAGES) ; do \
-		if [[ $$package != *minion/pb* ]] ; then \
+		if [[ $$package != *minion/pb* && $$package != *api/pb* ]] ; then \
```
Not an issue right now, but as we get more packages that we exclude from golint, we might consider doing something like what we do with `LINE_LENGTH_EXCLUDE` and `COV_SKIP`

mmussomele commented on api/server/server.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	var sock net.Listener
+	server := server{conn}
```
I don't this should have the same name as the struct type

kklin replied:
Fixed.

mmussomele commented on api/server/server.go:
```diff
@@ -0,0 +1,105 @@
+	for {
+		sock, err = net.Listen(proto, addr)
+
+		if err != nil {
```
```
if err == nil {
        break
}

log.WithError(err).Error("Failed to open socket.")
```
?

kklin replied:
Fixed.

mmussomele commented on api/server/server.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func (s server) QueryMachines(
+	cts context.Context, query *pb.DBQuery) (*pb.MachineReply, error) {
```
I would put a newline after this to visually separate the method header from the body.

kklin replied:
Fixed.

mmussomele commented on api/server/server.go:
```diff
@@ -0,0 +1,105 @@
+
+	var resp pb.MachineReply
+	for _, m := range machines {
+		resp.Machines = append(resp.Machines, &pb.Machine{
```
Would it be worth having an `api/util` package for converting between pb structs and db structs?

kklin replied:
I don't think it'd be that useful, the conversions only happen once.

mmussomele commented on api/server/server.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func (s server) QueryContainers(
+	cts context.Context, query *pb.DBQuery) (*pb.ContainerReply, error) {
```
Same here with the newline

kklin replied:
Fixed.

mmussomele commented on minion/main.go:
```diff
@@ -2,8 +2,11 @@
 	"time"
 
+	"github.com/NetSys/quilt/api"
+	apiServer "github.com/NetSys/quilt/api/server"
```
Why rename this?

kklin replied:
The minion package contains a type named `server`

mmussomele commented on quiltctl/containerCommand.go:
```diff
@@ -0,0 +1,37 @@
+package main
```
Is there a reason not to put containerCommand and machineCommand in one file named command?

kklin replied:
Many more commands get added in the next couple patches (`get`, `run`, `stop`, `inspect`, `ssh`)

mmussomele commented on quiltctl/containerCommand.go:
```diff
@@ -0,0 +1,37 @@
+
+func (cCmd *containerCommand) getFlagSet() *flag.FlagSet {
+	flags := flag.NewFlagSet("containers", flag.ExitOnError)
+
```
Why so many blank lines here?

kklin replied:
Personal preference, I guess. First part creates the flagSet, middle sets the flag mappings, final part returns.

mmussomele commented on quiltctl/containerCommand.go:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	for _, c := range allContainers {
+		fmt.Printf("%v\n", c)
```
Why a print instead of a log?

kklin replied:
This is the output of the command, while the logs were errors. Plus, this will go to stdout.

mmussomele commented on quiltctl/machineCommand.go:
```diff
@@ -0,0 +1,37 @@
+package main
```
Same comments as containerCommand.go

mmussomele commented on quiltctl/quiltctl.go:
```diff
@@ -0,0 +1,71 @@
+
+	cmd.run()
+
+	os.Exit(0)
```
Why not just let main exit?

kklin replied:
Done.

mmussomele commented on quiltctl/quiltctl_test.go:
```diff
@@ -0,0 +1,58 @@
+func (tCmd *testCommand) run() {
+	if tCmd.arg != "val" {
+		tCmd.t.Errorf("Bad argument value for testCommand: expected %s, got %s",
+			"val", tCmd.arg)
```
Why format the "val" into the error?

kklin replied:
Copy and paste, and fast editing :P. Fixed.

kklin commented on quiltctl/error.go:
```diff
@@ -0,0 +1,19 @@
+
+// Utility functions for displaying common errors.
+
+func daemonConnectError(host string, err error) {
```
Fixed the method name, not sure if the commands should all go in the same file.

mmussomele commented on cluster/foreman.go:
```diff
@@ -109,6 +110,11 @@ func (fm *foreman) runOnce() {
 
 		connected := err == nil
 		if connected && !m.connected {
+			fm.conn.Transact(func(view db.Database) error {
```
What if the machine becomes disconnected later? Should this transact be below, with `m.connected = connected`?

mmussomele commented on quiltctl/getCommand.go:
```diff
@@ -0,0 +1,56 @@
+package main
```
If you put the commands in the same file, bring this one along.

mmussomele commented on quiltctl/getCommand.go:
```diff
@@ -0,0 +1,56 @@
+		os.Exit(1)
+	}
+
+	fmt.Println("Successfully installed import!")
```
Log?

kklin replied:
Some feedback for users is nice

mmussomele commented on quiltctl/getCommand.go:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	if err := stitch.GetSpec(spec); err != nil {
+		log.WithError(err).Errorf("Error getting import `%s`.\n", spec)
```
Why the `\n`?

mmussomele replied:
Why the `\n`?

kklin replied:
Ah, I was used to fmt.Printf requiring a newline. I do this in a lot of places, will fix.

mmussomele commented on quiltctl/inspectCommand.go:
```diff
@@ -0,0 +1,21 @@
+package main
```
Fold into the single command file.

mmussomele commented on quiltctl/runCommand.go:
```diff
@@ -0,0 +1,98 @@
+package main
```
Fold into command file

mmussomele commented on quiltctl/runCommand.go:
```diff
@@ -0,0 +1,98 @@
+
+type runCommand struct {
+	stitch string
+
```
Why this blank line?

mmussomele commented on quiltctl/runCommand.go:
```diff
@@ -0,0 +1,98 @@
+		daemonResponseError(err)
+	}
+
+	fmt.Println("Successfully started run!")
```
Log?

mmussomele commented on quiltctl/stopCommand.go:
```diff
@@ -0,0 +1,62 @@
+		daemonResponseError(err)
+	}
+
+	fmt.Printf("Successfully began stopping namespace `%s`!\n", namespace)
```
same here

mmussomele commented on quiltctl/getCommand.go:
```diff
@@ -0,0 +1,56 @@
+		if len(nonFlagArgs) > 0 {
+			spec = nonFlagArgs[0]
+		} else {
+			log.Error("No import specified!")
```
Why so excited? XD

kklin replied:
Heh, I use exclamation marks in a couple place.. Wonder if I should just change all of them to periods. I think I'd get annoyed after awhile.

kklin replied:
Yeah, I ditched them.

mmussomele commented on quiltctl/sshCommand.go:
```diff
@@ -0,0 +1,95 @@
+package main
```
fold into command file

mmussomele commented on quiltctl/sshCommand.go:
```diff
@@ -0,0 +1,95 @@
+
+func (sCmd *sshCommand) run() {
+	args := sCmd.flags.Args()
+	tgtMachine, err := strconv.Atoi(args[0])
```
Why not just name it `targetMachine`? Not clear what `tgt` means

mmussomele commented on quiltctl/sshCommand.go:
```diff
@@ -0,0 +1,95 @@
+		log.Errorf("Unable to find machine `%d`.\n", tgtMachine)
+		log.Error("Available machines:")
+		for _, m := range machines {
+			log.Errorf("%v\n", m)
```
It might be cleaner to join these into a single string and log once.

mmussomele commented:
Should some of these commits should be folded together? It seems like migrating all the quilt commands into quiltctl should be one patch, adding new commands should be another, etc

kklin commented:
I'm not sure how I'd break up the commits otherwise. I wanted the stitch patch to be separate because it's half moving to quiltctl, and half restructuring the way we compile and then evaluate. Also, I like having separate commits for new features, makes it easier to revert if need be.

mmussomele commented:
Also, since quilt is now a daemon, should it's log message be redirected to a file instead of stdout? Like in quilt.go, should we have `log.SetOutput(some file writer)`?

kklin commented:
I think that's a good idea, I'll make a trello card for it. Doesn't have to happen in this series though.

kklin commented:
Pushed changes, ready for another round.

kklin commented:
Moved quiltctl commands to their own package.
I made an empty test file for the `command` package so that the build would pass.. I didn't see anything really worth testing, although as the commands increase in complexity I'd expect them to have some tests..

mmussomele commented on quiltctl/command/container.go:
```diff
@@ -0,0 +1,40 @@
+
+// Container contains the options for querying containers.
+type Container struct {
+	host string
```
Since this is the only field that Container has, why not simply have `type Container string`?

mmussomele replied:
Or if you want the meaning of the field to remain, `type ContainerHost string`?

kklin replied:
The container and machine subcommands will eventually take more flags, e.g.
```
quiltctl machines --query "key=Role,value=Master" --columns "ID,PublicIP"
```
Besides, I don't think we lose much by leaving it as a struct.

mmussomele commented on quiltctl/command/machine.go:
```diff
@@ -0,0 +1,40 @@
+
+// Machine contains the options for querying machines.
+type Machine struct {
+	host string
```
Same here. `type MachineHost string`?

mmussomele commented on quiltctl/quiltctl.go:
```diff
@@ -0,0 +1,63 @@
+
+var commands = map[string]command.SubCommand{
+	"machines":   &command.Machine{},
+	"containers": &command.Container{},
```
If the structs from before become strings, these could be replaced with `new(command.MachineHost)` and `new(command.ContainerHost)`

mmussomele commented on quiltctl/quiltctl.go:
```diff
@@ -0,0 +1,63 @@
+	/* XXX: GRPC spews a lot of uselss log message so we tell to eat its logs.
+	 * Once we have more sophistcated logging support, we should enable the log
+	 * messages when in debug mode. */
+	grpclog.SetLogger(l_mod.New(ioutil.Discard, "", 0))
```
`quilt.go` has support for setting the log level from the command line. Perhaps the level parsing code could be moved to `util/log.go` and both quilt and quiltctl could use it.

kklin replied:
There's actually something kind of related to this that I think needs to be figured out, curious if you have any ideas.
The way the quiltctl code is setup right now is that each subcommand has its own flagSet; however, there's some flags that are common to multiple commands (e.g. `host`, or `logLevel`)
I think the command format should eventually be like:
`quiltctl -H tcp://8.8.8.8 machines --queryString "Role=Master"`

mmussomele replied:
logLevel should definitely be a flag of quiltctl, but I'm not sure about H. It doesn't make much sense to say something like:
`quiltctl -H tcp://8.8.8.8 stop test-namespace` but that would be valid if H was a quiltctl flag right?

mmussomele replied:
As we add more commands that require the host flag, we could wrap them further in a QueryCommand interface or something if things were getting repetitive.

ejj replied:
-H is what docker uses fwiw

Ethan (Phone)

> On Jul 31, 2016, at 21:01, mmussomele <notifications@github.com> wrote:
> 
> In quiltctl/quiltctl.go:
> 
> > +	"github.com/NetSys/quilt/quiltctl/command"
> > +
> > +	log "github.com/Sirupsen/logrus"
> > +	"google.golang.org/grpc/grpclog"
> > +)
> > +
> > +var commands = map[string]command.SubCommand{
> > +	"machines":   &command.Machine{},
> > +	"containers": &command.Container{},
> > +}
> > +
> > +func main() {
> > +	/* XXX: GRPC spews a lot of uselss log message so we tell to eat its logs.
> > +	 * Once we have more sophistcated logging support, we should enable the log
> > +	 * messages when in debug mode. */
> > +	grpclog.SetLogger(l_mod.New(ioutil.Discard, "", 0))
> logLevel should definitely be a flag of quiltctl, but I'm not sure about H. It doesn't make much sense to say something like:
> quiltctl -H tcp://8.8.8.8 stop test-namespace but that would be valid if H was a quiltctl flag right?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mmussomele commented on quiltctl/command/get.go:
```diff
@@ -0,0 +1,59 @@
+		} else {
+			log.Error("No import specified.")
+			gCmd.usage()
+		}
```
```
if len(nonFlagArgs) == 0 {
        log.Error("No import specified.")
        gCmd.usage()
}
spec = nonFlagArgs[0]
```

mmussomele replied:
`gCmd.usage` exits, so you can reduce indentation here

kklin replied:
Thanks, I'll make the same change for `run` and `stop`.

mmussomele commented on quiltctl/command/run.go:
```diff
@@ -0,0 +1,100 @@
+		} else {
+			log.Error("No spec specified.")
+			rCmd.usage()
+		}
```
Same here with reworking the if-else into just an if

kklin commented:
Pushed changes to reduce if-indentation.

sirspinach commented on quiltctl/quiltctl.go:
```diff
@@ -0,0 +1,63 @@
+func parseSubcommand(name string, args []string) (command.SubCommand, error) {
+	cmd, ok := commands[name]
+	if !ok {
+		return nil, fmt.Errorf("unrecongized subcommand: %s", name)
```
unrecognized is misspelled. (haha this is my contribution?)

sirspinach commented on stitch/stitch.go:
```diff
@@ -108,6 +97,38 @@ func New(sc scanner.Scanner, path string, download bool) (Stitch, error) {
 	return spec, nil
 }
 
+// Compile takes a Stitch (possible with imports), and transforms it into a
```
possibly with?

kklin commented:
Fixed typos pointed out by shwang.

ejj commented on api/client/client.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+// QueryMachines retrieves the machines tracked by the Quilt daemon.
+func (c Client) QueryMachines() ([]db.Machine, error) {
```
I think you could unit test these pretty darn easily if you mock out pbClient . . .  Not hugely valuable, but I am trying to get the code coverage up.

ejj commented on api/client/client.go:
```diff
@@ -0,0 +1,94 @@
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+	reply, err := c.pbClient.QueryContainers(ctx, &pb.DBQuery{})
+	if err != nil {
+		return []db.Container{}, err
```
I think you can just `return nil, err`

ejj commented on api/common_test.go:
```diff
@@ -0,0 +1,43 @@
+func TestAddrParsing(t *testing.T) {
+	t.Parallel()
+
+	checkAddrResult :=
```
This is kinda ugly, just make it a regular function in the top level?  I think you'd just need to add a `t *testing.T` argument.

kklin commented:
@mmussomele: Updated the first commit to generalize the DB querying and return responses as JSON, could you review it when you have time?

mmussomele commented:
Sure, I'll probably get around to it tonight

ejj commented on api/client/client.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+// Client provides methods to interact with the Quilt daemon.
+type Client struct {
```
Can you do `type Client pb.APIClient`?

kklin replied:
No, `pb.APIClient` is an interface.

ejj commented on api/client/client.go:
```diff
@@ -0,0 +1,85 @@
+		return containers, nil
+	}
+
+	panic(fmt.Sprintf("unsupported table type: %s", table))
```
I'd slightly prefer this be the default case of the switch statement.  Tad more clear IMO.

ejj commented on api/common.go:
```diff
@@ -0,0 +1,24 @@
+)
+
+// DefaultSocket is the socket the Quilt daemon listens on by default.
+const DefaultSocket = "unix:///tmp/quilt.sock"
```
/tmp is a weird place for this, where does docker put it's socket?  OVS puts it in the ovs equivalent of `/var/run/quilt/quilt.sock`

kklin replied:
You're right, but /var/run is owned by root it looks like.. We could run the daemon as root, but then the client will also have to be run by root to write to the socket. We could circumvent this with some permissions twiddling, but I'm not sure it's worth putting in the effort before we think through what's most usable.

ejj commented on api/server/server.go:
```diff
@@ -0,0 +1,83 @@
+		time.Sleep(30 * time.Second)
+	}
+
+	// Cleanup the socket if we're interrupted.
```
Huh . . . what happens if we don't do this?  I'm surprised we have to do something so wonky

kklin replied:
I had it there because the server would fail to create the socket the next time it started up. Doing it this way does give us locking (no more than one instance of quilt/machine) for free, though. I don't feel strongly either way.

ejj commented on api/server/server.go:
```diff
@@ -0,0 +1,83 @@
+	var rows interface{}
+	switch db.TableType(query.Table) {
+	case db.MachineTable:
+		s.dbConn.Transact(func(view db.Database) error {
```
Not sure it matters all that much, but it might be *slightly* cleaner if you put the switch inside the transaction, instead of building a transaction within each switch.  The other nice thing is that as we add more tables it will be simpler for those that don't have a SelectFrom implemented on their conn.

ejj commented:
First and Second patches looks good.

ejj commented:
Third patch looks good.

ejj commented:
Fourth Patch looks good.

kklin commented on Makefile:
```diff
@@ -20,14 +20,19 @@ install:
 
 check: format-check
-	go test $(PACKAGES)
+	# XXX: Hack to make `make check` skip the quilt-tester test scripts.
```
@ejj: wasn't sure how to deal with this, so I just threw this here to pass the build.

kklin commented:
Updated the PR with:
- More coverage in `quiltctl/command`. The `Command` interface changed a bit to facilitate this. Also cut down on the number of random `os.Exit`s.
- Updates to quilt-tester
    - Switched over to using quiltctl
    - Changed the code structure a bit, moving a couple things into structs and improving the error handling.
    - Changed the test suite structure to get rid of the `src` subdirectories
    - Made tests run in a namespace based on the machine's IP.

ejj commented on api/server/server.go:
```diff
@@ -0,0 +1,105 @@
+		case db.ContainerTable:
+			rows = view.SelectFromContainer(nil)
+		default:
+			foundTable = false
```
Why not just return `fmt.Errorf(.....)`?  Then you can ditch this foundTable variable and check check the error that Transact returns.

ejj commented on quilt-tester/config/infrastructure.spec:
```diff
@@ -0,0 +1,12 @@
+// We will have three worker machines.
+(define nWorker 3)
+
+(define Namespace "tester-52-53-247-146")
```
Do you think it's hard to fix this?  I'd really like to before we merge, it's just going to cause problems.

kklin replied:
It's fixed, I can change the spec so that it's clear.

ejj commented on quilt-tester/exec.go:
```diff
@@ -0,0 +1,245 @@
+// waitFor waits until `pred` is satisfied, or `timeout` seconds have passed.
+func waitFor(pred func() bool, timeout int) error {
+	timeoutChan := time.After(time.Duration(timeout) * time.Second)
+	for range time.Tick(1 * time.Second) {
```
Meh, just `for {` and `time.Sleep`  Simpler, doesn't leak the ticker.

kklin replied:
How does that work with the timeout?

ejj replied:
You could still use the time.After chan and check on each loop.  I'm just saying the time.Tick isn't really providing any value.

ejj commented on quilt-tester/exec.go:
```diff
@@ -0,0 +1,245 @@
+		return "", "", err
+	}
+
+	cmdFinished := make(chan error, 1)
```
Huh.  I don't quite get the point of this `cmdFinished` thing.  Why can't you just call `cmd.Wait()`?

kklin replied:
Totally can...

ejj commented on quilt-tester/exec.go:
```diff
@@ -0,0 +1,245 @@
+func cleanupMachines(namespace string) {
+	l := log.testerLogger
+
+	l.infoln("Cleaning up first with `quilt stop`.")
```
Same question, why the oddness with the logging.

kklin commented on quilt-tester/exec.go:
```diff
@@ -0,0 +1,245 @@
+func execCmd(cmd *exec.Cmd, logLineTitle string) (string, string, error) {
+	l := log.cmdLogger
+
+	l.infoln(fmt.Sprintf("%s: Starting command: %v", logLineTitle, cmd.Args))
```
We have the `fileLogger`s to save logs to the files on the website. Are you saying we should use logrus loggers instead of rolling our own?

ejj replied:
Yeah, if it's not too hard.  If it's a big project, then don't bother.

kklin replied:
It'd take more effort than I'd like to dedicate to the tester at this point.

kklin commented:
- Addressed previous round comments
- Added a commit to close the grpc connection (sigh)
- Fixed a bug where the quilt-tester pass/fail counts didn't update
- Added some unit tests for quilt-tester
